sphinx/source/docs/dev_guide/testing.rst

```diff
@@ -1,231 +1,447 @@
+  For example: ``pytest -n 4 tests/codebase``. See `pytest-xdist`_.
+* ``-v``: Run test with more verbose output.
+* ``--driver``: Use a specific web driver for Selenium-based tests
+  (``"chrome"``, ``"firefox"``, or ``"safari"``). For example:
+  ``pytest --driver="firefox" tests/integration/``.
+* ``--no-js``: Skip any JavaScript code and only test Python code.
 
-    pytest -m "not selenium" tests/unit
+See the `pytest documentation`_ for more options.
 
-To run just the BokehJS unit tests, execute:
+Codebase tests
+    The most basic set of tests are Bokeh's codebase tests. This includes
+    linting with `flake8` and other tests for issues such as unused imports and
+    extra white spaces.
 
-.. code-block:: sh
+    Run this command from the top level of the repository:
 
-    pytest tests/test_bokehjs.py
+    .. code-block:: sh
 
-Alternatively, you can also navigate to the `bokehjs` subdirectory of the
-source checkout and execute:
+        pytest tests/codebase
 
-.. code-block:: sh
+Unit tests
+    To run Bokeh's Python unit tests, use the following command at the top
+    level of the repository:
 
-    node make test
+    .. code-block:: sh
 
-You can run all available tests (python and JS unit tests, as well as example
-and integration tests) **from the top-level directory** by executing:
+        pytest -m "not selenium" tests/unit
 
-.. code-block:: sh
+    .. note::
+        This command will exclude unit tests that require `Selenium`_. Because
+        Selenium can be difficult to set up and because some unit tests require
+        both `geckodriver`_ and `ChromeDriver`_ to be available on your system,
+        using ``-m "not selenium"`` is the recommended way to run unit tests
+        locally. Once you :ref:`create a Pull Request <devguide_pull_requests>`,
+        :ref:`Bokeh's CI <devguide_testing_ci>` will run all tests, including
+        Selenium-based unit tests. In case Selenium with both geckodriver and
+        ChromeDriver is available on your system, you can run all unit tests
+        with ``pytest tests/unit``.
 
-    pytest
+.. _devguide_testing_local_python_integration:
 
-To learn more about marking test functions and selecting/deselecting them for
-a run, please consult the pytest documentation for `custom markers`_. The list
-of currently defined test markers is below:
+Integration tests
+    To run Bokeh's Python-focused integration tests, use this command from the
+    top level of the repository:
 
-* ``sampledata``: a test for requiring ``bokeh.sampledata`` be downloaded
-* ``selenium``: a test requiring selenium
+    .. code-block:: sh
 
-Code Coverage
-~~~~~~~~~~~~~
+        pytest tests/integration
 
-To run any of the tests with coverage, use the following:
+    These tests mostly simulate UI interactions. Therefore, they require
+    `Chrome`_ or `Chromium`_ and `Selenium`_ with the `ChromeDriver`_ web
+    driver.
 
-.. code-block:: sh
+Code coverage
```

Comment:
@bryevdv @mattpap I was not able to find many details about coverage in the current docs. Do we have a certain coverage (x%) that we are aiming for? Or is this mostly just for informational purposes?

Reply:
Coverage is tricky since we have lots of different kinds of tests and the coverage tool(s) only work with some of them The instructions here are only for the Python unit tests, which hover around 90% and I would like to keep it that way. There is some way to also get JS unit test coverage but I don't remember what it is. We've never looked at it with any regularity. The would be two completely separate reports which is also a pain. In principle I think the different coverage tools can all output XML reports that  could potentially be merged in a post-processing step, but no one has ever looked into it. For the selenium, examples, visual js tests, etc. I'm not sure if it is possible to get coverage info at all.

TLDR; Wnat ot keep Python unit test coverage around 90+% we should keep information about that if nothing else.

Reply:
Sounds good! I updated the section to clarify that this only applies to Python unit tests and that the goal is around 90% coverage there.